code/espurna/tuya.cpp

```diff
@@ -197,57 +349,63 @@ namespace Tuya {
         }
 
         if (State::DISCOVERY == state) {
-            discoveryTimeout.feed();
-            pushOrUpdateState(type, frame);
-        } else {
+            discovery.add(type, frame[0]);
+        } else if (!filter) {
             updateState(type, frame);
         }
-
     }
 
     void processFrame(State& state, const Transport& buffer) {
 
-        const DataFrame frame(buffer);
+        const DataFrameView frame(buffer);
 
         dataframeDebugSend("<=", frame);
 
         // initial packet has 0, do the initial setup
         // all after that have 1. might be a good idea to re-do the setup when that happens on boot
-        if (frame.commandEquals(Command::Heartbeat) && (frame.length == 1)) {
-            if (State::HEARTBEAT == state) {
-                if ((frame[0] == 0) || !configDone) {
-                    DEBUG_MSG_P(PSTR("[TUYA] Starting configuration ...\n"));
+        if ((frame.command() == Command::Heartbeat) && (frame.length() == 1)) {
+            if (State::BOOT == state) {
+                if (!configDone) {
+                    DEBUG_MSG_P(PSTR("[TUYA] Attempting to configure the board ...\n"));
+#if LIGHT_PROVIDER == LIGHT_PROVIDER_CUSTOM
+                    setupChannels();
+#endif
+                    setupSwitches();
+                }
+
+                if (!configDone) {
+                    DEBUG_MSG_P(PSTR("[TUYA] Starting discovery\n"));
```

Comment:
As mentioned, now this waits until serial receives at least one heartbeat message (sending every ~200 ... ~400 ... ~600 ms and etc.)